modules/javafx.graphics/src/main/native-glass/win/KeyTable.cpp

```diff
@@ -343,21 +343,46 @@ BOOL IsExtendedKey(UINT vkey) {
 }
 
 /*
- * Class:     Java_com_sun_glass_ui_win_WinApplication
- * Method:    _getKeyCodeForChar
- * Signature: (C)I
+ * Class:     com_sun_glass_ui_win_WinApplication
+ * Method:    _getKeyCanGenerateCharacter
+ * Signature: (IIC)Z
  */
-JNIEXPORT jint JNICALL Java_com_sun_glass_ui_win_WinApplication__1getKeyCodeForChar
-  (JNIEnv * env, jobject jApplication, jchar c)
+JNIEXPORT jboolean JNICALL Java_com_sun_glass_ui_win_WinApplication__1getKeyCanGenerateCharacter
+  (JNIEnv *env, jobject jApplication, jint hardwareCode, jint vkCode, jchar character)
 {
-    BYTE vkey = 0xFF & ::VkKeyScanEx((TCHAR)c,
-            ::GetKeyboardLayout(GlassApplication::GetMainThreadId()));
+    HKL layout = ::GetKeyboardLayout(GlassApplication::GetMainThreadId());
+
+    // Normally Windows doesn't generate a character for the Delete key but the
+    // ViewContainer synthesizes one
+    if (vkCode == com_sun_glass_events_KeyEvent_VK_DELETE) {
+        return character == 0x7F;
+    }
+
+    // This will match keys on the numeric keypad as well as characters on the
+    // main keyboard that don't require modifiers
+    if (hardwareCode >= 0) {
+        UINT keyChar = ::MapVirtualKeyEx(UINT(hardwareCode), 2, layout);
+        // Filter out dead keys
+        BOOL isDead = (keyChar & 0x80000000) != 0;
```

Comment:
Wouldn't `hardwareCode >= 0` already filter out dead keys since when the highest bit is set the value is negative?

Reply:
I just realized that I never actually said what `hardwareCode` means on this platform.

A negative number means the `hardwareCode` isn't known (in other words, the KeyEvent wasn't constructed by glass). If it's not negative it's a Windows VK code that identifies the key the user pressed. That code doesn't reflect whether the key is dead or not. Here I'm asking what character that key would generate if pressed without any modifiers and ignoring it if the key is dead.

Reply:
I misread partially, somehow I read `keyChar` as the same as `hardwareCode`. LGTM then.